docs/architecture/decisions/0014-dynamic-datasource-registration.md

```diff
@@ -0,0 +1,106 @@
+# 0014. Dynamic Data Source Registration
+
+## Status
+
+Accepted
+
+## Context
+
+Decouple the data sources from the SDK into their own nuget packages and create a mechanism to allow auto discovery and dynamic registration. This will allow for new data sources to be plugged in with minimum configuration and dependency on the existing code base. 
+
+## Decision
+
+- Decoupling from SDK
+
+Each data source can be built externally with no dependency on the CarbonAware SDK. This helps achieve single responsibility principle and makes the system extensible.
+
+- Increased Maintainability
+
+Currently, the data source registration code is implemented using a switch statement to register the appropriate data source based on configuration. This can be removed completely and only the data source needed can be imported into the SDK.
+
+- Controlled Solution Size
+
+All the data sources are part of the current SDK, even if they are not used at runtime. For e.g., if both **IEmissions** and **IForecast** interfaces are configured to use WattTime, we need not package the JSON and ElectricityMaps data sources, thereby reducing solution’s size.
+
+- Reduction of Security threads
+
+Having decoupled data sources as packages, it allows us to control potential security threads that can be injected into the overall system, by providing the opportunity to certify them.
+
+## Consequences
+
+Be able to have outsourced Data Sources would benefit the overall system in case there is a large number of those by allowing which one should be part of a solution.
+
+## Design Considerations
+
+Currently Data Sources project is consumed by GSF library using Dependency Injection. Each GSF library handlers have references to IForecastDataSource and/or IEmissionsDataSource Data Source interfaces.
+To modify this interaction and to make it more dynamic, these are things that required to be considered:
+
+- Public interfaces & Data Records
+
+    Data Sources interfaces and Data Records are internal and available only to certain projects (i.e GSF.CarbonAware). This would require to be changed so consumers can dynamically register those and consume them.
+
+    Effort Level: **Medium**
+
+- Packaging
+
+    Current Data Source project is not allowed to be packaged (nuget package), which would require how this is going to be done in terms of, what the package contains, versioning and where to publish it.
```

Comment:
Can you add a few more words or an additional sentence with further context about why packaging is not allowed for someone that doesn't know? A short explanation would suffice.